File: lcd_bbb.py
import Adafruit_BBIO.GPIO as GPIO
import time

# LCD GPIO pin mapping
LCD_RS = "P8_7"
LCD_RW = "P8_8"    # Tie to GND if using write-only
LCD_EN = "P8_9"
LCD_D4 = "P8_10"
LCD_D5 = "P8_11"
LCD_D6 = "P8_12"
LCD_D7 = "P8_14"

# Setup pins as output
pins = [LCD_RS, LCD_EN, LCD_D4, LCD_D5, LCD_D6, LCD_D7]
for pin in pins:
    GPIO.setup(pin, GPIO.OUT)
GPIO.setup(LCD_RW, GPIO.OUT)
GPIO.output(LCD_RW, GPIO.LOW)   # Always Write mode

def pulse_e():
    GPIO.output(LCD_EN, GPIO.HIGH)
    time.sleep(0.001)
    GPIO.output(LCD_EN, GPIO.LOW)
    time.sleep(0.001)

def send_4bits(bits):
    GPIO.output(LCD_D4, GPIO.HIGH if bits & 0x1 else GPIO.LOW)
    GPIO.output(LCD_D5, GPIO.HIGH if bits & 0x2 else GPIO.LOW)
    GPIO.output(LCD_D6, GPIO.HIGH if bits & 0x4 else GPIO.LOW)
    GPIO.output(LCD_D7, GPIO.HIGH if bits & 0x8 else GPIO.LOW)
    pulse_e()

def send_byte(data, mode):
    GPIO.output(LCD_RS, GPIO.HIGH if mode else GPIO.LOW)
    send_4bits((data >> 4) & 0x0F)
    send_4bits(data & 0x0F)
    time.sleep(0.001)

def lcd_init():
    send_4bits(0x03)
    time.sleep(0.005)
    send_4bits(0x03)
    time.sleep(0.002)
    send_4bits(0x03)
    send_4bits(0x02)         # Set 4-bit mode

    send_byte(0x28, 0)       # 2 line, 5x8 matrix
    send_byte(0x0C, 0)       # Display on, cursor off
    send_byte(0x06, 0)       # Entry mode set
    send_byte(0x01, 0)       # Clear display
    time.sleep(0.002)

def lcd_write(msg):
    for char in msg:
        send_byte(ord(char), 1)

def lcd_goto(row, col):
    # LCD rows: 0 or 1
    addr = col + (0x40 if row else 0x00)
    send_byte(0x80 | addr, 0)

if __name__ == "__main__":
    lcd_init()
    lcd_write("Hello, BBB!")
    lcd_goto(1,0)
    lcd_write("16x2 LCD Demo")

File: lcd_bbb.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <unistd.h>
#include <cstdlib>

using namespace std;

#define LCD_RS 66   // Example: P8_7 → GPIO66
#define LCD_RW 67   // Example: P8_8 → GPIO67
#define LCD_EN 69   // Example: P8_9 → GPIO69
#define LCD_D4 68   // Example: P8_10 → GPIO68
#define LCD_D5 45   // Example: P8_11 → GPIO45
#define LCD_D6 44   // Example: P8_12 → GPIO44
#define LCD_D7 26   // Example: P8_14 → GPIO26


void gpio_export(int pin) {
    ofstream exportFile("/sys/class/gpio/export");
    exportFile << pin;
    exportFile.close();
    usleep(100000);
}

// Set direction
void gpio_direction(int pin, string dir) {
    ofstream dirFile("/sys/class/gpio/gpio" + to_string(pin) + "/direction");
    dirFile << dir;
    dirFile.close();
}

// Write value
void gpio_write(int pin, int value) {
    ofstream valFile("/sys/class/gpio/gpio" + to_string(pin) + "/value");
    valFile << value;
    valFile.close();
}

// Pulse enable pin
void pulse_e() {
    gpio_write(LCD_EN, 1);
    usleep(1000);
    gpio_write(LCD_EN, 0);
    usleep(1000);
}

// Send 4 bits
void send_4bits(int bits) {
    gpio_write(LCD_D4, (bits & 0x1) ? 1 : 0);
    gpio_write(LCD_D5, (bits & 0x2) ? 1 : 0);
    gpio_write(LCD_D6, (bits & 0x4) ? 1 : 0);
    gpio_write(LCD_D7, (bits & 0x8) ? 1 : 0);
    pulse_e();
}

// Send byte to LCD
void send_byte(int data, bool mode) {
    gpio_write(LCD_RS, mode ? 1 : 0);
    send_4bits((data >> 4) & 0x0F);
    send_4bits(data & 0x0F);
    usleep(1000);
}

// LCD init sequence
void lcd_init() {
    send_4bits(0x03);
    usleep(5000);
    send_4bits(0x03);
    usleep(2000);
    send_4bits(0x03);
    send_4bits(0x02);   // 4-bit mode

    send_byte(0x28, false);  // 2 line, 5x8 font
    send_byte(0x0C, false);  // Display ON, cursor OFF
    send_byte(0x06, false);  // Increment cursor
    send_byte(0x01, false);  // Clear display
    usleep(2000);
}

// Write string to LCD
void lcd_write(string msg) {
    for (char c : msg) {
        send_byte((int)c, true);
    }
}

// Move cursor
void lcd_goto(int row, int col) {
    int addr = col + (row ? 0x40 : 0x00);
    send_byte(0x80 | addr, false);
}

int main() {
    // Export pins
    int pins[] = {LCD_RS, LCD_RW, LCD_EN, LCD_D4, LCD_D5, LCD_D6, LCD_D7};
    for (int pin : pins) {
        gpio_export(pin);
        gpio_direction(pin, "out");
    }

    // Always write mode
    gpio_write(LCD_RW, 0);

    lcd_init();
    lcd_write("Hello, BBB!");
    lcd_goto(1, 0);
    lcd_write("16x2 LCD Demo");

    return 0;
}
